### PR TITLE
Gra 1054/egress

### PR DESCRIPTION
**Pull request by abhishek9686**
Testing Steps:
server should point to https://github.com/gravitl/netmaker/pull/1998

- [x] create an egress gateway on a linux node
- [x] check if other peers are able to reach the machines in the egress ranges
- [x] check with ext clients too if they are able to reach the egress ranges

### CODE REVIEW

**Review by mattkasun — CHANGES_REQUESTED** [review]
fix failing integration tests

File: nmproxy/router/firewall_nonlinux.go [hunk 1]
```diff
@@ -40,6 +40,17 @@ func (unimplementedFirewall) FlushAll() {
 
 }
 
+func (unimplementedFirewall) InsertEgressRoutingRules(server string, egressInfo models.EgressInfo) error {
```
Comment (0xdcarns) [inline nmproxy/router/firewall_nonlinux.go:43]: missing comments  on functions

File: nmproxy/router/iptables_linux.go [hunk 2]
```diff
@@ -361,6 +358,121 @@ func (i *iptablesManager) InsertIngressRoutingRules(server string, extinfo model
...
+			}
+		}
+
+	}
+	ruleTable[egressInfo.EgressID].rulesMap[egressInfo.EgressID] = egressGwRoutes
+
+	return nil
+}
+
+func (i *iptablesManager) AddEgressRoutingRule(server string, egressInfo models.EgressInfo,
```
Comment (0xdcarns) [inline nmproxy/router/iptables_linux.go:442]: missing comment

File: nmproxy/router/iptables_linux.go [hunk 3]
```diff
@@ -384,9 +496,25 @@ func (i *iptablesManager) FetchRuleTable(server string, tableName string) ruleta
...
 		}
+	case egressTable:
+		rules = i.engressRules[server]
+		if rules == nil {
+			rules = make(ruletable)
+		}
 	}
 	return rules
 }
+func (i *iptablesManager) DeleteRuleTable(server, ruleTableName string) {
```
Comment (0xdcarns) [inline nmproxy/router/iptables_linux.go:507]: missing comment

**Review by 0xdcarns — CHANGES_REQUESTED** [review]

**Comment by theguy951357** [comment]
will wait on changes

File: nmproxy/router/firewall_nonlinux.go (hunk 1 above)
Comment (abhishek9686) [reply nmproxy/router/firewall_nonlinux.go:43]: it's unimplemented firewall funcs, so comments on these funcs are not required

**Review by 0xdcarns — APPROVED** [review]

**Review by mattkasun — CHANGES_REQUESTED** [review]

**Review by mattkasun — CHANGES_REQUESTED** [review]
fix staticcheck failures

**Comment by theguy951357** [comment]
will wait on changes

**Comment by theguy951357** [comment]
setup branches on server and nodes.
joined ubuntuQA and fedoraQA.
initial pings to eachother work.
made an egress gateway on ubuntuQA with VPC range 10.116.0.0/20
should be able to ping rockylinuxQA on same network with a vpc on 10.116.0.13
fedoraQA cannot ping 10.116.0.13

**Comment by abhishek9686** [comment]
> setup branches on server and nodes. joined ubuntuQA and fedoraQA. initial pings to eachother work. made an egress gateway on ubuntuQA with VPC range 10.116.0.0/20 should be able to ping rockylinuxQA on same network with a vpc on 10.116.0.13 fedoraQA cannot ping 10.116.0.13

fixed the issue you can re-check it now

File: functions/mqhandlers.go [hunk 4]
```diff
@@ -159,6 +159,8 @@ func HostPeerUpdate(client mqtt.Client, msg mqtt.Message) {
 	}
 	config.UpdateHostPeers(serverName, peerUpdate.Peers)
 	config.WriteNetclientConfig()
+	nc := wireguard.NewNCIface(config.Netclient(), config.GetNodes())
```
[the remark this answers is not in the document]
Comment (abhishek9686) [reply functions/mqhandlers.go:162]: the interface is not getting re-created, it's fetching new addresses and setting routes

**Review by mattkasun — CHANGES_REQUESTED** [review]
staticchecks

File: nmproxy/router/firewall_linux.go [hunk 5]
```diff
@@ -34,3 +39,18 @@ func isIptablesSupported() bool {
 	_, err6 := exec.LookPath("ip6tables")
 	return err4 == nil && err6 == nil
 }
+
+func getInterfaceName(dst net.IPNet) (string, error) {
+	h, err := netlink.NewHandle(0)
+	if err != nil {
+		return "", err
+	}
+	routes, err := h.RouteGet(dst.IP)
```
Comment (mattkasun) [inline nmproxy/router/firewall_linux.go:48]: err is not handled

**Review by mattkasun — CHANGES_REQUESTED** [review]

**Review by theguy951357 — APPROVED** [review]
egress range is reachable.
interfaces loook normal.
ext clients can ping egress range.
pass

**Pull request merged**